Psubscribe support

janoc opened the pull request:
Here is a basic PSUBSCRIBE/PUNSUBSCRIBE support

I have done some basic tests, seems to work, but more testing is welcome. I am also not 100% sure about the public API - I have added usePattern boolean as the last (and optional) argument to the existing subscribe/unsubscribe functions in order to avoid code duplication, but perhaps it would be better to create separate functions for it.

nekipelov commented:
I want to add methods for all redis command later. Thats why we should keep `subscribe` semantics. Can you move `subscribe` implementation to the RedisClientImpl and add `psubscribe`?

Will be like this:

```cpp
RedisAsyncClient::subscribe(channel, messagesHandler, handler)
{
    return pimpl->subscribe("subscribe", channel, messagesHandler, handler);
}

RedisAsyncClient::psubscribe(channel, messagesHandler, handler)
{
    return pimpl->subscribe("psubscribe", channel, messagesHandler, handler);
}
```

janoc commented:
OK, I have moved the code to the implementation class and added separate functions. It required a bit of because the Handle struct is defined in the RedisAsyncClient and not available in the impl class.

nekipelov commented on src/redisclient/impl/redisclientimpl.cpp:
```diff
@@ -56,12 +56,14 @@ void RedisClientImpl::doProcessMessage(RedisValue v)
     if( state == State::Subscribed )
     {
         std::vector<RedisValue> result = v.toArray();
+        auto rsize = result.size();
```
Please rename `rsize` to `resultSize`.

nekipelov commented on src/redisclient/impl/redisasyncclient.cpp:
```diff
@@ -83,123 +83,42 @@ RedisAsyncClient::Handle RedisAsyncClient::subscribe(
+    std::function<void(std::vector<char> msg)> msgHandler,
+    std::function<void(RedisValue)> handler)
+{
+    auto handle_id = pimpl->subscribe("psubscribe", pattern, msgHandler, handler);
```
Please rename `handle_id` to `handleId`.

nekipelov commented on src/redisclient/impl/redisasyncclient.cpp:
```diff
@@ -83,123 +83,42 @@ RedisAsyncClient::Handle RedisAsyncClient::subscribe(
-        ss << "RedisAsyncClient::command called with invalid state "
-           << to_string(pimpl->state);
+    auto handle_id = pimpl->subscribe("subscribe", channel, msgHandler, handler);    
+    return { handle_id , channel };
```
Please rename `handle_id` to `handleId`.

nekipelov commented:
It looks good. Can you rename some variables and remove commit e60eec030e1d6d28c3e88d11eef640f4066894fb (rebase and squash)?

janoc commented:
I have renamed the variables as you wanted. However, I didn't remove that commit, because the branch would stop building. 

Either cherrypick the commits you want or we need to first fix and merge #32 and then rebase this on it. It will save me the work of having to fix the build twice.

janoc commented:
Nope, it doesn't build, because of the duplicate symbols found (the autolinking issue). Will do a temporary hack on my side that will not be included in the pull req. to get it to build until #32 is merged.

janoc commented:
OK, done - AppVeyor CI may not build because of the double linking in Windows, that is to be expected.

nekipelov commented:
Excellent. Thank you!
